File: go/problems/pg42578/solution.go
// Package pg42578 - [PG] 42578 - 의상
// https://programmers.co.kr/learn/courses/30/lessons/42578
// 난이도: lv2
// 태그: hash
//
// 시간복잡도: O(n)
// 공간복잡도: O(n)
package pg42578

func Solution(clothes [][]string) int {
	// 최종 크기를 예상할 . 있다면, 반드시 초기 용량을 설정- 메모리 할당 비용이 크게 감소
	wardrobe := make(map[string]int, len(clothes))
	// key, value로 정리
	for i := range clothes {
		key := clothes[i][1]
		wardrobe[key]++
	}
	result := 1
	for _, value := range wardrobe {
		result *= value + 1
	}

	return result - 1
}

File: java/src/main/java/problems/pg42578/Solution.java
package problems.pg42578;

import java.util.Arrays;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * [PG] 42578 - 의상 https://programmers.co.kr/learn/courses/30/lessons/42578 난이도: lv2 태그: hash
 *
 * <p>시간복잡도: O(n) 공간복잡도: O(n)
 */
public class Solution {

  public int solution(String[][] clothes) {
    final Closet closet = new Closet(clothes);
    return closet.getCount();
  }

  private static class Closet {
    private final Map<String, Long> closet;

    public Closet(final String[][] clothes) {

      this.closet =
          Arrays.stream(clothes)
              .collect(Collectors.groupingBy(cloth -> cloth[1], Collectors.counting()));
    }

    public int getCount() {
      //      reduce(초기값, (누적값, 현재값) -> 계산식)
      return Math.toIntExact(closet.values().stream().map(i -> i + 1).reduce(1L, (a, b) -> (a * b)) - 1);
    }
  }
}
